src/datasets/arrow_dataset.py

```diff
@@ -3348,13 +3402,22 @@ def add_column(self, name: str, column: Union[list, np.array], new_fingerprint:
             :class:`Dataset`
         """
         column_table = InMemoryTable.from_pydict({name: column})
+        _check_column_names(self.column_names + column_table.column_names)
+        # Pad tables if needed to match the maximum length
+        max_num_rows = max(self.num_rows, column_table.num_rows)
+        dset_table = pad_table_to_length(self._data, max_num_rows)
+        column_table = pad_table_to_length(column_table, max_num_rows)
+        indices_table = self._indices
+        if indices_table is not None and max_num_rows > self.num_rows:
+            indices_to_add = range(self._data.num_rows, self._data.num_rows + (max_num_rows - self.num_rows))
+            indices_table = add_indices_to_indices_table(self._indices, indices_to_add)
```

Comment:
Is this really necessary ? It could lead to more frequent silent errors rather than helping users no ?

Reply:
Actually, I agree with you that this could lead to silent errors. Also, the code is definitely not the nicest :). So let's disable manual padding in `add_column`/`concatenate_datasets` for now, and only allow padding introduced by `pa.concat_tables(..., promote=True)`!